DOC Update the user guide entry for Nystroem kernel approximation (#27446)
scikit-learn/scikit-learn: doc/modules/kernel_approximation.rst

:math:`U_2 \Lambda U_2^T`. To do this we can express it in terms of the already
evaluated matrices:

.. math::

         \begin{align} U_2 \Lambda U_2^T &= \left(K_{21} U_1 \Lambda^{-1}\right) \Lambda \left(K_{21} U_1 \Lambda^{-1}\right)^T
         \\&= K_{21} U_1 (\Lambda^{-1} \Lambda) \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} U_1 \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} K_{11}^{-1} K_{21}^T
         \\&= \left( K_{21} K_{11}^{-\frac12} \right) \left( K_{21} K_{11}^{-\frac12} \right)^T 
         .\end{align}

During ``fit``, the class :class:`Nystroem` evaluates the basis :math:`U_1`, and
computes the normalization constant, :math:`K_{11}^{-\frac12}`. Later, during
``transform``, the kernel matrix is determined between the basis (given by the
`components_` attribute) and the new data points, ``X``. This matrix is then
multiplied by the ``normalization_`` matrix for the final result.

By default :class:`Nystroem` uses the ``rbf`` kernel, but it can use any kernel
function or a precomputed kernel matrix. The number of samples used - which is
also the dimensionality of the features computed - is given by the parameter

> is given by the parameter ``n_components``.

Reviewer: ```suggestion
``n_components``.

.. topic:: Examples:

    * :ref:`sphx_glr_auto_examples_applications_plot_cyclical_feature_engineering.py`
```

Nystroem Method for Kernel Approximation
----------------------------------------
The Nystroem method, as implemented in :class:`Nystroem` is a general method for
reduced rank approximations of kernels. It achieves this by subsampling without
replacement rows/columns of the data on which the kernel is evaluated. While the
computational complexity of the exact method is
:math:`\mathcal{O}(n^3_{\text{samples}})`, the complexity of the approximation
is :math:`\mathcal{O}(n^2_{\text{components}} \cdot n_{\text{samples}})`, where
one can set :math:`n_{\text{components}} \ll n_{\text{samples}}` without a
significative decrease in performance [WS2001]_.

> For the dataset we can construct eigendecomposition of kernel matrix `K` and
> split it by sampled and unsampled data

Reviewer: ```suggestion
We can construct the eigendecomposition of the kernel matrix :math:`K`, based
on the features of the data, and then split it into sampled and unsampled data
points.
```

.. math::

        K = U \Lambda U^T
        = \begin{bmatrix} U_1 \\ U_2\end{bmatrix} \Lambda \begin{bmatrix} U_1 \\ U_2 \end{bmatrix}^T
        = \begin{bmatrix} U_1 \Lambda U_1^T & U_1 \Lambda U_2^T \\ U_2 \Lambda U_1^T & U_2 \Lambda U_2^T \end{bmatrix}
        \equiv \begin{bmatrix} K_{11} & K_{12} \\ K_{21} & K_{22} \end{bmatrix}

where:

    * :math:`U` is orthonormal
    * :math:`Ʌ` is diagonal matrix of eigenvalues
    * :math:`U_1` is orthonormal matrix of samples that were chosen
    * :math:`U_2` is orthonormal matrix of samples that were not chosen

Given that :math:`U_1 \Lambda U_1^T` can be obtained by orthonormalization of
the matrix :math:`K_{11}`, and :math:`U_2 \Lambda U_1^T` can be evaluated (as

> well as its transpose), the only remaining term to elucidate is 

Reviewer: ```suggestion
well as its transpose), the only remaining term to elucidate is
```

.. math::

        K = U \Lambda U^T
        = \begin{bmatrix} U_1 \\ U_2\end{bmatrix} \Lambda \begin{bmatrix} U_1 \\ U_2 \end{bmatrix}^T
        = \begin{bmatrix} U_1 \Lambda U_1^T & U_1 \Lambda U_2^T \\ U_2 \Lambda U_1^T & U_2 \Lambda U_2^T \end{bmatrix}
        \equiv \begin{bmatrix} K_{11} & K_{12} \\ K_{21} & K_{22} \end{bmatrix}

where:

    * :math:`U` is orthonormal
    * :math:`Ʌ` is diagonal matrix of eigenvalues
    * :math:`U_1` is orthonormal matrix of samples that were chosen
    * :math:`U_2` is orthonormal matrix of samples that were not chosen

Given that :math:`U_1 \Lambda U_1^T` can be obtained by orthonormalization of
the matrix :math:`K_{11}`, and :math:`U_2 \Lambda U_1^T` can be evaluated (as
well as its transpose), the only remaining term to elucidate is
:math:`U_2 \Lambda U_2^T`. To do this we can express it in terms of the already
evaluated matrices:

.. math::

         \begin{align} U_2 \Lambda U_2^T &= \left(K_{21} U_1 \Lambda^{-1}\right) \Lambda \left(K_{21} U_1 \Lambda^{-1}\right)^T
         \\&= K_{21} U_1 (\Lambda^{-1} \Lambda) \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} U_1 \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} K_{11}^{-1} K_{21}^T

>          \\&= \left( K_{21} K_{11}^{-\frac12} \right) \left( K_{21} K_{11}^{-\frac12} \right)^T 

Reviewer 2: ```suggestion
         \\&= \left( K_{21} K_{11}^{-\frac12} \right) \left( K_{21} K_{11}^{-\frac12} \right)^T
```

         .\end{align}

During ``fit``, the class :class:`Nystroem` evaluates the basis :math:`U_1`, and
computes the normalization constant, :math:`K_{11}^{-\frac12}`. Later, during
``transform``, the kernel matrix is determined between the basis (given by the
`components_` attribute) and the new data points, ``X``. This matrix is then
multiplied by the ``normalization_`` matrix for the final result.

By default :class:`Nystroem` uses the ``rbf`` kernel, but it can use any kernel
function or a precomputed kernel matrix. The number of samples used - which is
also the dimensionality of the features computed - is given by the parameter
``n_components``.

.. topic:: Examples:

>     * :ref:`sphx_glr_auto_examples_applications_plot_cyclical_feature_engineering.py`

Reviewer 2: ```suggestion
    * See the example entitled
      :ref:`sphx_glr_auto_examples_applications_plot_cyclical_feature_engineering.py`,
      that shows a efficient machine learning pipeline that uses a
      :class:`Nystroem` kernel.
```

Nystroem Method for Kernel Approximation
----------------------------------------

> The Nystroem method, as implemented in :class:`Nystroem` is a general method

Reviewer: ```suggestion
The Nystroem method, as implemented in :class:`Nystroem` is a general method for
```

for reduced rank approximations of kernels. It achieves this by subsampling without replacement
rows/columns of the data on which the kernel is evaluated.

> While the exact computational complexity is :math:\mathcal{O}(n^3_{\text{samples}}),

Reviewer 2: ```suggestion
While the exact computational complexity is :math:`\mathcal{O}(n^3_{\text{samples}})`,
```

> one can set :math:n_{\text{components}} \ll n_{\text{samples}} without a significative decrease in performance 

Reviewer 2: ```suggestion
one can set :math:`n_{\text{components}} \ll n_{\text{samples}}` without a significative
decrease in performance 
```

> for reduced rank approximations of kernels. It achieves this by subsampling without replacement
> rows/columns of the data on which the kernel is evaluated.
> While the exact computational complexity is :math:`\mathcal{O}(n^3_{\text{samples}})`,
> one can set :math:`n_{\text{components}} \ll n_{\text{samples}}` without a significative decrease in performance 
> (:math:`\mathcal{O}(n^2_{\text{components}} \cdot n_{\text{samples}})`) [WS2001]_.
> 
> For the dataset we can construct eigendecomposition of kernel matrix `K` and split it by sampled and unsampled data

Reviewer: ```suggestion
reduced rank approximations of kernels. It achieves this by subsampling without
replacement rows/columns of the data on which the kernel is evaluated. While the
computational complexity of the exact method is
:math:`\mathcal{O}(n^3_{\text{samples}})`, the complexity of the approximation
is :math:`\mathcal{O}(n^2_{\text{components}} \cdot n_{\text{samples}})`, where
one can set :math:`n_{\text{components}} \ll n_{\text{samples}}` without a
significative decrease in performance [WS2001]_.

For the dataset we can construct eigendecomposition of kernel matrix `K` and
split it by sampled and unsampled data
```

.. math::

>         K = U \Lambda U^T
>         = \begin{bmatrix} U_1 \\ U_2\end{bmatrix} \Lambda \begin{bmatrix} U_1 \\ U_2 \end{bmatrix}^T
>         = \begin{bmatrix} U_1 \Lambda U_1^T & U_1 \Lambda U_2^T \\ U_2 \Lambda U_1^T & U_2 \Lambda U_2^T \end{bmatrix}

Reviewer: ```suggestion
        K = U \Lambda U^T
        = \begin{bmatrix} U_1 \\ U_2\end{bmatrix} \Lambda \begin{bmatrix} U_1 \\ U_2 \end{bmatrix}^T
        = \begin{bmatrix} U_1 \Lambda U_1^T & U_1 \Lambda U_2^T \\ U_2 \Lambda U_1^T & U_2 \Lambda U_2^T \end{bmatrix}
        \equiv \begin{bmatrix} K_{11} & K_{12} \\ K_{21} & K_{22} \end{bmatrix}
```

where:

>     * :math:``U`` is orthonormal

Reviewer 2: ```suggestion
    * :math:`U` is orthonormal
```

>     * :math:``Ʌ`` is diagonal matrix of eigenvalues

Reviewer 2: ```suggestion
    * :math:`Ʌ` is diagonal matrix of eigenvalues
```

    * :math:`U_1` is orthonormal matrix of samples that were chosen
    * :math:`U_2` is orthonormal matrix of samples that were not chosen

> Since we already evaluated :math:`U_1 \Lambda U_1^T` by orthonormalization of matrix `K`, and can evaluate :math:`U_2 \Lambda U_1^T`.
> The only thing for us to find out is :math:`U_2 \Lambda U_2^T`.
> To do this we can try to express it from the already evaluated matricies

Reviewer: ```suggestion
Given that :math:`U_1 \Lambda U_1^T` can be obtained by orthonormalization of
the matrix :math:`K_{11}`, and :math:`U_2 \Lambda U_1^T` can be evaluated (as
well as its transpose), the only remaining term to elucidate is 
:math:`U_2 \Lambda U_2^T`. To do this we can express it in terms of the already
evaluated matrices:
```

.. math::

         \begin{align} U_2 \Lambda U_2^T &= \left(K_{21} U_1 \Lambda^{-1}\right) \Lambda \left(K_{21} U_1 \Lambda^{-1}\right)^T
         \\&= K_{21} U_1 (\Lambda^{-1} \Lambda) \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} U_1 \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} K_{11}^{-1} K_{21}^T
         \\&= \left( K_{21} K_{11}^{-\frac12} \right) \left( K_{21} K_{11}^{-\frac12} \right)^T 
         .\end{align}

During ``fit``, the class :class:`Nystroem` evaluates the basis :math:`U_1`, and
computes the normalization constant, :math:`K_{11}^{-\frac12}`. Later, during
``transform``, the kernel matrix is determined between the basis (given by the

> components_ attribute) and the new datapoints, ``X``. This matrix is then

Reviewer: This was being interpreted as a link
```suggestion
`components_` attribute) and the new data points, ``X``. This matrix is then
```

multiplied by the ``normalization_`` matrix for the final result.

> By default :class:`Nystroem` uses the ``rbf`` kernel, but it can use any
> kernel function or a precomputed kernel matrix.
> The number of samples used - which is also the dimensionality of the features computed -

Reviewer: ```suggestion
By default :class:`Nystroem` uses the ``rbf`` kernel, but it can use any kernel
function or a precomputed kernel matrix. The number of samples used - which is
also the dimensionality of the features computed - is given by the parameter
```

Nystroem Method for Kernel Approximation
----------------------------------------
The Nystroem method, as implemented in :class:`Nystroem` is a general method
for reduced rank approximations of kernels. It achieves this by subsampling without replacement
rows/columns of the data on which the kernel is evaluated.

> Computational complexity for this method is :math:`\mathcal{O}(n^2_{\text{components}} \cdot n_{\text{samples}})` [WS2001]_

Reviewer: Whereas the exact kernel calculation is is :math:`\mathcal{O}(n^3_{\text{samples}})`.
One can set :math:`n_{\text{components}} \ll n_{\text{samples}}` without a significative decrease in performance.

.. math::

        K = U \Lambda U^T
        = \begin{bmatrix} U_1 \\ U_2\end{bmatrix} \Lambda \begin{bmatrix} U_1 \\ U_2 \end{bmatrix}^T
        = \begin{bmatrix} U_1 \Lambda U_1^T & U_1 \Lambda U_2^T \\ U_2 \Lambda U_1^T & U_2 \Lambda U_2^T \end{bmatrix}

where:

    * ``U`` is orthonormal
    * ``Ʌ`` is diagonal matrix of eigenvalues
    * :math:`U_1` is orthonormal matrix of samples that were chosen
    * :math:`U_2` is orthonormal matrix of samples that were not chosen

We already know :math:`U_1 \Lambda U_1^T`, and can evaluate :math:`U_2 \Lambda U_1^T`.
The only thing for us to find out is :math:`U_2 \Lambda U_2^T`.
To do this we can try to express it from the already evaluated matricies

.. math::

         \begin{align}
         U_2 \Lambda U_2^T
         \\&= \left(K_{21} U_1 \Lambda^{-1}\right) \Lambda \left(K_{21} U_1 \Lambda^{-1}\right)^T
         \\&= K_{21} U_1 (\Lambda^{-1} \Lambda) \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} U_1 \Lambda^{-1} U_1^T K_{21}^T
         \\&= K_{21} K_{11}^{-1} K_{21}^T
         \\&= \left( K_{21} K_{11}^{-\frac12} \right) \left( K_{21} K_{11}^{-\frac12} \right)^T 
         .\end{align}

> We will need to calculate kernel from :math:`U_1` to :math:`U_2` (:math:`K_{21}`)
> and then normalize it by :math:`K_{11}^{-\frac12}`.
> During the ``fit`` method :class:`Nystroem` will evaluate basis (:math:`U_1`)
> and normalization constant (:math:`K_{11}^{-\frac12}`).
> During ``transform`` method kernel matrix between basis (``components_`` attribute) and new
> datapoints (``X``) will be evaluated. After that it will be multiplied it by the ``normalization_`` matrix.

Reviewer: ```suggestion
During ``fit``, the class :class:`Nystroem` evaluates the basis :math:`U_1`, and
computes the normalization constant,:math:`K_{11}^{-\frac12}`. Later, during
``transform``, the kernel matrix is determined between the basis (given by the
components_ attribute) and the new datapoints, ``X``. This matrix is then
multiplied by the ``normalization_`` matrix for the final result.
```
